overlord/hookstate/ctlcmd/model.go

```diff
@@ -0,0 +1,265 @@
+	snapInfo.Publisher, err = assertstate.PublisherStoreAccount(st, snapInfo.SnapID)
+	return snapInfo, err
+}
+
+// checkGadgetOrModel verifies that the current snap context is either a gadget snap or that
+// the snap shares the same publisher as the current model assertion.
+func (c *modelCommand) checkGadgetOrModel(st *state.State, deviceCtx snapstate.DeviceContext, snapInfo *snap.Info) error {
+	// We allow the usage of this command if one of the following is true
+	// 1. The requesting snap must be a gadget
+	// 2. Come from the same brand as the device model assertion
+	// 3. Have the snapd-control plug
+	if snapType := snapInfo.Type(); snapType == snap.TypeGadget {
+		return nil
+	}
+	if snapInfo.Publisher.ID == deviceCtx.Model().BrandID() {
+		return nil
+	}
+	if hasSnapdControlInterface(st, snapInfo.SnapName()) {
+		return nil
+	}
+
+	c.reportError("cannot get model assertion for snap %q: not a gadget or from the same brand as the device model assertion\n",
+		snapInfo.SnapName())
+	return fmt.Errorf("insufficient permissions to get model assertion for snap %q", snapInfo.SnapName())
+}
+
+// findSerialAssertion is a helper function to find the newest matching serial assertion
+// for the provided model assertion.
+func findSerialAssertion(st *state.State, modelAssertion *asserts.Model) *asserts.Serial {
+	assertions, err := assertstate.DB(st).FindMany(asserts.SerialType, map[string]string{
+		"brand-id": modelAssertion.BrandID(),
+		"model":    modelAssertion.Model(),
+	})
+	if err != nil || len(assertions) == 0 {
+		return nil
+	}
+
+	// helper to parse the timestamp embedded in the assertion, if
+	// anything goes wrong parsing this we return an empty timestamp
+	// to only use it in case there is no other serial assertion with
+	// a valid timetamp
+	getAssertionTime := func(a asserts.Assertion) time.Time {
+		iTimestamp := a.HeaderString("timestamp")
+
+		// for assertions the timestamp is always RFC3339
+		t1, err := time.Parse(time.RFC3339, iTimestamp)
+		if err != nil {
+			return time.Time{}
+		}
+		return t1
+	}
+
+	sort.Slice(assertions, func(i, j int) bool {
+		t1 := getAssertionTime(assertions[i])
+		t2 := getAssertionTime(assertions[j])
+		return t1.Before(t2)
+	})
+
+	// sort.Slice sorts in ascending order, so the last assertion is the newest
+	serial := assertions[len(assertions)-1].(*asserts.Serial)
+	return serial
+}
+
+func (c *modelCommand) Execute([]string) error {
+	context, err := c.ensureContext()
+	if err != nil {
+		return err
+	}
+
+	// ignore the valid bool as we just pass the task whether it is
+	// nil or not.
+	task, _ := context.Task()
+	st := context.State()
+	st.Lock()
+	defer st.Unlock()
+
+	deviceCtx, err := snapstate.DeviceCtx(st, task, nil)
+	if err != nil {
+		return err
+	}
+
+	// We only return an error in case we could not the get the snap.Info
+	// structure, and 'ignore' any error that caused us not to get the store
+	// account publisher
+	snapInfo, err := getSnapInfo(st, context.InstanceName())
```

Comment:
The only case where getSnapInfo() returns both an error and a snapInfo != nil (which is intentional) is if it fails to find the snap declaration/publisher account in the database, and as far as I understand we want to continue in this case, since the snap declaration/store account is actually optional for retreiving the model assertion. (Again, as far as I understand, I may very well be wrong). The flow here is purely intentional.